src/java/org/apache/cassandra/io/sstable/format/SSTableReader.java

```diff
@@ -808,32 +809,12 @@ public void runOnClose(final Runnable runOnClose)
         synchronized (tidy.global)
         {
             final Runnable existing = tidy.runOnClose;
-            tidy.runOnClose = AndThen.get(existing, runOnClose);
-        }
-    }
-
-    private static class AndThen implements Runnable
-    {
-        final Runnable runFirst;
-        final Runnable runSecond;
-
-        private AndThen(Runnable runFirst, Runnable runSecond)
-        {
-            this.runFirst = runFirst;
-            this.runSecond = runSecond;
-        }
-
-        public void run()
-        {
-            runFirst.run();
-            runSecond.run();
-        }
-
-        static Runnable get(Runnable runFirst, Runnable runSecond)
-        {
-            if (runFirst == null)
-                return runSecond;
-            return new AndThen(runFirst, runSecond);
+            tidy.runOnClose = () -> {
```

Comment:
AFAIR `AndThen` here is necessary because a lambda retains a reference to the reader, which we definitely don't want (see comment on top of the method).

If you revert this change, please add a "do not replace with a lambda..." comment in front of `AndThen` itself.

Reply:
I must disagree. To my knowledge lambda does not include a reference to the enclosing instance unless explicitly `this` or some member is used. I've verified that, also see this answer https://stackoverflow.com/a/35636556

Reply:
[This](https://issues.apache.org/jira/browse/CASSANDRA-9423?focusedCommentId=14602771&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-14602771) is the original problem stemming from this. AFAIR Stefania later fought the same problem with lambdas (where `this` was included in the closure even when it was not required), but I may be wrong, or things may have changed.

If you have verified that leak detection works correctly for all supported JDK versions, I'll be very happy to accept that this is one less problem we have to worry about.

Reply:
added test cases in `RefCountedTest`

Reply:
Great.

Could you add comments explaining that lambdas, if certain that they don't reference `this`, are okay?

As a nit, `existing` can be checked immediately, and `tidy.runOnClose` can be set to the argument directly if it's null.

Reply:
fee2c251d370fd61ca436bc16169c1d1774ec01c

Reply:
The new tests look great. Thank you for adding them.